lib/monitor.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2002-2018 Balabit
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "monitor.h"
+
+#if SYSLOG_NG_HAVE_INOTIFY
+static void
+handler(void *_fm, struct inotify_event *event)
+{
+  FileMonitor *fm = (FileMonitor *) _fm;
+
+  if (event->mask & IN_MODIFY)
+    fm->main_loop_reload_config(fm->self);
+
+  // 1- handle case where editing the file using vim causes an event 'IN_IGNORED'
+  //     and subsequently the watch gets deleted by ivykis: ivykis/src/iv_inotify.c:88
+  //   2- handle case when file is edited and contains a parse error, syslog-ng changes
```

Comment:
I do not think parsing error need special handling, just simply trigger the reload (and that if the config fails fall back if can to the previous config).
I would ignore the case when because of attribute change the config is not possible to read anymore. Most of the time syslog-ng is running as root anyway (or with DAC_OVERRIDE cap by default). If it turns out that is a needed feature, probably somebody is going to ask for it, and it won't be hard to do that time.

Reply:
The special handling is not related to inotify itself, but the way syslog-ng - for some reason - modifies the attributes of the configuration file if a reload has been issued with a problematic file, and then inotify is not able to watch it anymore because it's marked as 'deleted'. I don't know the root cause for that but I will reproduce it and confirm to double check.